### PR TITLE
feat(controller): Allow to configure main container resources

### PR DESCRIPTION
**Pull request by terrytangyuan**
Some clusters require explicit resources in all containers. This would allow us to configure the resources for the main container.

Signed-off-by: terrytangyuan <terrytangyuan@gmail.com>

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

**Comment by alexec** [comment]
podSpecPatch?

**Comment by terrytangyuan** [comment]
Thanks for the suggestion. It's definitely an option. However, podSpecPatch requires all of our existing users to modify their workflow spec which is not trivial (especially when we have long-running workflows and established usage patterns). Configuring this in the config of controller seems to be a better option for now.

File: workflow/controller/workflowpod.go [hunk 1]
```diff
@@ -151,6 +151,9 @@ func (woc *wfOperationCtx) createWorkflowPod(nodeName string, mainCtr apiv1.Cont
 	wfSpec := woc.execWf.Spec.DeepCopy()
 
 	mainCtr.Name = common.MainContainerName
+	if isResourcesSpecified(woc.controller.Config.MainContainer) {
+		mainCtr.Resources = woc.controller.Config.MainContainer.Resources
```
Comment (alexec) [inline workflow/controller/workflowpod.go:155]: 1. The workflow spec should take priority over the config.
2. Test?

Comment (terrytangyuan) [reply workflow/controller/workflowpod.go:157]: Thanks for the feedback. Just updated the PR.

Comment (alexec) [reply workflow/controller/workflowpod.go:157]: I think you can ignore `podSpePatch`, but you should check tmpl.Container

Comment (terrytangyuan) [reply workflow/controller/workflowpod.go:157]: Could you clarify why we need to check tmpl.Container? The goal of this PR is to customize main container no matter what's in the template (similar to what's currently done for executor).

Comment (alexec) [reply workflow/controller/workflowpod.go:157]: that would be contrary to how we do it for other item - the user should be able to override a global setting like this

Comment (terrytangyuan) [reply workflow/controller/workflowpod.go:158]: Actually another goal of this PR is to set reasonable default resources for the main container.

I've added a check on `tmpl.Container`. Let me know if I've missed anything here.

**Review by alexec — APPROVED** [review]

**Pull request merged**